store/sqlstore/scheme_store.go

```diff
@@ -293,66 +305,105 @@ func (s *SqlSchemeStore) Delete(schemeId string) (*model.Scheme, error) {
 	s.Channel().ClearCaches()
 
 	// Delete the roles belonging to the scheme.
+
+	inQuery := sq.Or{}
 	roleNames := []string{scheme.DefaultChannelGuestRole, scheme.DefaultChannelUserRole, scheme.DefaultChannelAdminRole}
-	if scheme.Scope == model.SchemeScopeTeam {
-		roleNames = append(roleNames, scheme.DefaultTeamGuestRole, scheme.DefaultTeamUserRole, scheme.DefaultTeamAdminRole)
+
+	for _, ele := range roleNames {
+		inQuery = append(inQuery, sq.Eq{"Name": ele})
 	}
+	if scheme.Scope == model.SchemeScopeTeam {
+		additionalRoleNames := []string{scheme.DefaultTeamGuestRole, scheme.DefaultTeamUserRole, scheme.DefaultTeamAdminRole}
 
-	var inQueryList []string
-	queryArgs := make(map[string]interface{})
-	for i, roleId := range roleNames {
-		inQueryList = append(inQueryList, fmt.Sprintf(":RoleName%v", i))
-		queryArgs[fmt.Sprintf("RoleName%v", i)] = roleId
+		for _, ele := range additionalRoleNames {
+			inQuery = append(inQuery, sq.Eq{"Name": ele})
+		}
 	}
-	inQuery := strings.Join(inQueryList, ", ")
 
 	time := model.GetMillis()
-	queryArgs["UpdateAt"] = time
-	queryArgs["DeleteAt"] = time
 
-	if _, err := s.GetMaster().Exec("UPDATE Roles SET UpdateAt = :UpdateAt, DeleteAt = :DeleteAt WHERE Name IN ("+inQuery+")", queryArgs); err != nil {
+	updateQuery, args, err := s.getQueryBuilder().
+		Update("Roles").
+		Where(inQuery).
+		Set("UpdateAt", time).
+		Set("DeleteAt", time).
+		ToSql()
+
+	if err != nil {
+		return nil, errors.Wrap(err, "status_tosql")
+	}
+
+	if _, err = s.GetMasterX().Exec(updateQuery, args...); err != nil {
 		return nil, errors.Wrapf(err, "failed to update Roles with name in (%s)", inQuery)
 	}
 
 	// Delete the scheme itself.
 	scheme.UpdateAt = time
 	scheme.DeleteAt = time
 
-	rowsChanged, err := s.GetMaster().Update(&scheme)
+	res, err := s.GetMasterX().NamedExec(`UPDATE Schemes
+		SET UpdateAt=:UpdateAt, DeleteAt=:DeleteAt, CreateAt=:CreateAt, Name=:Name, DisplayName=:DisplayName, Description=:Description, Scope=:Scope,
+		 DefaultTeamAdminRole=:DefaultTeamAdminRole, DefaultTeamUserRole=:DefaultTeamUserRole, DefaultTeamGuestRole=:DefaultTeamGuestRole,
+		 DefaultChannelAdminRole=:DefaultChannelAdminRole, DefaultChannelUserRole=:DefaultChannelUserRole, DefaultChannelGuestRole=:DefaultChannelGuestRole 
+		 WHERE Id=:Id`, &scheme)
+
 	if err != nil {
 		return nil, errors.Wrapf(err, "failed to update Scheme with schemeId=%s", schemeId)
 	}
+
+	rowsChanged, err := res.RowsAffected()
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get RowsAffected while updating scheme with schemeId=%s", schemeId)
+	}
 	if rowsChanged != 1 {
 		return nil, errors.New("no record to update")
 	}
 	return &scheme, nil
 }
 
 func (s *SqlSchemeStore) GetAllPage(scope string, offset int, limit int) ([]*model.Scheme, error) {
-	var schemes []*model.Scheme
+	schemes := []*model.Scheme{}
+
+	scopeQuery := sq.And{}
+	scopeQuery = append(scopeQuery, sq.Eq{"DeleteAt": 0})
```

Comment:
We can keep the `DeleteAt = 0` part in the original query itself. Just don't call `ToSql` immediately. Move the `if scope != ""` part after that, and call `Where(sq.Eq{"Scope": scope}) on the query and re-assign the variable`. Squirrel takes care of it automatically. And then call `ToSql` before executing the query.

This makes the overall flow more readable.

Reply:
Great insights @agnivade. I have included your changes and I can see the code is more readable. Learnt something more today :100: ! Thank You!